### PR TITLE
Solution Assignment 4

### PR DESCRIPTION
**Pull request by Utena**
* This commit update the repository with a recursive and non-recursive solution (disjoint sets) of "Assignment 4" (main task).
* Optional challenge is still missing.

### CODE REVIEW

File: Utena/assignment-4/DisjointSet.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * This class implements a disjoint set to compute the number of islands on a 2D map. 
+ * 
+ * @author Jordina Torrents-Barrena
+ * @version 1.0
+ */
+public class DisjointSet {
+
+	private int[] parent, rank;
```
Comment (NickFortescueGoogle) [inline Utena/assignment-4/DisjointSet.java]: The google style guide (and I think the Oracle one) recommends just having one variable declared per line:
https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations

There are a few reasons for this:
- clarity, it is easier to see the type of each variable
- change logs. If the variables change in a pull request, it is really quick to see which one changed.

File: Utena/assignment-4/DisjointSet.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	
+	public DisjointSet(int numSquareTiles) {
+	    parent = rank = new int[numSquareTiles + 1];
```
Comment (NickFortescueGoogle) [inline Utena/assignment-4/DisjointSet.java]: I also think that for clarity I would normally expand this to multiple assignments. You've got to remember that not everyone reading code will know the language as well as you. So writing:

rank = new int[numSquareTiles + 1];
parent = rank;

is much clearer to read in the order of assignments. And it makes clear that parent and rank share the same array (which seems a bit odd to be honest). I'm very surprised the code works when they share an array. Is it well tested?

When you add a class like DisjointSet it is good to add unit tests for this class as well. That's why they are called "unit" tests. They test one unit at a time. Could you add tests for this class on its own?

File: Utena/assignment-4/DisjointSet.java (hunk 1 above)
Comment (Utena) [reply Utena/assignment-4/DisjointSet.java]: Yes, I agree! I've done the changes.

File: Utena/assignment-4/DisjointSet.java (hunk 2 above)
Comment (Utena) [reply Utena/assignment-4/DisjointSet.java]: Yes, you're right. It is better to expant this line to multiple assignments. 

In the case of "parent = rank", I would say that both arrays share the same dimensions. I change the code accordingly to clarify that point:

rank = new int[numSquareTiles + 1];
parent = new int[numSquareTiles + 1];

Moreover, I also add JUnit tests for the class DisjointSet.java!

**Comment by Utena** [comment]
Assignment 4: Revision 1 has been done.

**Comment by Utena** [comment]
Done!

**Comment by Utena** [comment]
I've done all requested reviews, but I cannot merge the changes to my branch!

**Pull request closed without merging**